### ISSUE TITLE
Canvas Grid Flickering (partially fixed)

### ISSUE THREAD

**Issue by iamtaeuk** [CigarProject/Cigar#385]
Hello @MastaCoder ,

From the recent Chrome updates, the grid on the canvas starts to flicker. IE and Edge is fine, but Chrome and Safari are the browsers that are affected by this. I have attached a screenshot, and you can see it flickering on my server [here](http://beta.agario.kr/berlin).

![image](https://cloud.githubusercontent.com/assets/13702277/20453172/ae4163c0-ae5f-11e6-99c3-47592c678117.png)

Thanks :)

**Comment by HAlexTM** [comment]
I can confirm, this happens on my grandma's PC that has a higher resolution than 1366x768, what's your screen resolution?

**Comment by tveloper** [comment]
@HAlexTM The non-beta server does not use the new Cigar client, it has the old client files. The beta server that I just got has the newest Cigar client. My screen resolution is FHD (1920x1080). 

Affected on Chrome, Safari, and other Chromium (Opera etc.) browsers.

**Comment by MastaCoder** [comment]
I can confirm this works on mine as well, if people are okay with me removing the grid lines all together I can do that.

**Comment by tveloper** [comment]
@MastaCoder How about a checkbox to turn off gridlines?

**Comment by MastaCoder** [comment]
Honestly I think the grids just need to go all together. They are pretty much broken throughout the browsers and are just terribly made. (Performance issues as well)

Until we can find a way to get the new one that Agar.io uses :(

**Comment by HAlexTM** [comment]
@MastaCoder I made different tests with, without and with different methods grids:
![onpaste 20161122-134048](https://cloud.githubusercontent.com/assets/11839341/20524172/741f0d4e-b0b9-11e6-89f8-9ede55b45537.png)
![onpaste 20161122-134124](https://cloud.githubusercontent.com/assets/11839341/20524184/8a27775c-b0b9-11e6-8f14-d0503b17c6b1.png)
![onpaste 20161122-134155](https://cloud.githubusercontent.com/assets/11839341/20524175/78537076-b0b9-11e6-8099-8adfe9aee69b.png)

There is no performance problems here... I'm pretty much confused

**Comment by Gimmo** [comment]
Problem is scaling. If display hardware resolution not equal to resolution set in adapter, it always have glitches of this kind.

**Comment by HAlexTM** [comment]
>  (Performance issues as well)
I’m talking about this

**Comment by HAlexTM** [comment]
@MastaCoder
> Until we can find a way to get the new one that Agar.io uses :(

agar.io/js/master.js
````
grid: {
        fillStyle: "#000000",
        strokeStyle: "#777777",
        lineWidth: 1,
        sharpLines: false,
        millisPerLine: 1e3,
        verticalSections: 2,
        borderVisible: true
     },
//...//
    context.fillStyle = chartOptions.grid.fillStyle;
        context.clearRect(0, 0, dimensions.width, dimensions.height);
        context.fillRect(0, 0, dimensions.width, dimensions.height);
        context.restore();
        context.save();
        context.lineWidth = chartOptions.grid.lineWidth;
        context.strokeStyle = chartOptions.grid.strokeStyle;
        if (chartOptions.grid.millisPerLine > 0) {
            context.beginPath();
            for (var t = time - time % chartOptions.grid.millisPerLine; t >= oldestValidTime; t -= chartOptions.grid.millisPerLine) {
                var gx = timeToXPixel(t);
                if (chartOptions.grid.sharpLines) {
                    gx -= .5
                }
                context.moveTo(gx, 0);
                context.lineTo(gx, dimensions.height)
            }
            context.stroke();
            context.closePath()
        }
        for (var v = 1; v < chartOptions.grid.verticalSections; v++) {
            var gy = Math.round(v * dimensions.height / chartOptions.grid.verticalSections);
            if (chartOptions.grid.sharpLines) {
                gy -= .5
            }
            context.beginPath();
            context.moveTo(0, gy);
            context.lineTo(dimensions.width, gy);
            context.stroke();
            context.closePath()
        }
        if (chartOptions.grid.borderVisible) {
            context.beginPath();
            context.strokeRect(0, 0, dimensions.width, dimensions.height);
            context.closePath()
        }
        context.restore();
        if (chartOptions.horizontalLines && chartOptions.horizontalLines.length) {
            for (var hl = 0; hl < chartOptions.horizontalLines.length; hl++) {
                var line = chartOptions.horizontalLines[hl],
                    hly = Math.round(valueToYPixel(line.value)) - .5;
                context.strokeStyle = line.color || "#ffffff";
                context.lineWidth = line.lineWidth || 1;
                context.beginPath();
                context.moveTo(0, hly);
                context.lineTo(dimensions.width, hly);
                context.stroke();
                context.closePath()
            }
        }
        for (var d = 0; d < this.seriesSet.length; d++) {
            context.save();
            var timeSeries = this.seriesSet[d].timeSeries,
                dataSet = timeSeries.data,
                seriesOptions = this.seriesSet[d].options;
            timeSeries.dropOldData(oldestValidTime, chartOptions.maxDataSetLength);
            context.lineWidth = seriesOptions.lineWidth;
            context.strokeStyle = seriesOptions.strokeStyle;
            context.beginPath();
            var firstX = 0,
                lastX = 0,
                lastY = 0;
            for (var i = 0; i < dataSet.length && dataSet.length !== 1; i++) {
                var x = timeToXPixel(dataSet[i][0]),
                    y = valueToYPixel(dataSet[i][1]);
                if (i === 0) {
                    firstX = x;
                    context.moveTo(x, y)
                } else {
                    switch (chartOptions.interpolation) {
                        case "linear":
                        case "line":
                            {
                                context.lineTo(x, y);
                                break
                            }
                        case "bezier":
                        default:
                            {
                                context.bezierCurveTo(Math.round((lastX + x) / 2), lastY, Math.round(lastX + x) / 2, y, x, y);
                                break
                            }
                        case "step":
                            {
                                context.lineTo(x, lastY);
                                context.lineTo(x, y);
                                break
                            }
                    }
                }
                lastX = x;
                lastY = y
            }
            if (dataSet.length > 1) {
                if (seriesOptions.fillStyle) {
                    context.lineTo(dimensions.width + seriesOptions.lineWidth + 1, lastY);
                    context.lineTo(dimensions.width + seriesOptions.lineWidth + 1, dimensions.height + seriesOptions.lineWidth + 1);
                    context.lineTo(firstX, dimensions.height + seriesOptions.lineWidth);
                    context.fillStyle = seriesOptions.fillStyle;
                    context.fill()
                }
                if (seriesOptions.strokeStyle && seriesOptions.strokeStyle !== "none") {
                    context.stroke()
                }
                context.closePath()
            }
            context.restore()
        }

**Comment by MastaCoder** [comment]
@HAlexTM Eh maybe performance doesn't get affected that much. Where did you get the clean Agar.io script from? 😆

**Comment by HAlexTM** [comment]
Well.. from agar.io ahah ([view-source:](view-source:http://agar.io))

**Comment by CalvinDKL** [comment]
I remember that i fixed this completely on my old client, but i dont really use the grid.
But ill look into it :)

**Comment by MastaCoder** [comment]
@CalvinDKL If you have anything send it to us 👍

**Comment by AuroraYT** [comment]
Hey thats easy to fix Calvin just do :)

    function drawGrid() {
        ctx.fillStyle = showDarkTheme ? "#111111" : "#F2FBFF";
        ctx.fillRect(0, 0, canvasWidth, canvasHeight);
        ctx.save();
        ctx.strokeStyle = showDarkTheme ? "#AAAAAA" : "#000000";
        ctx.globalAlpha = .2;
        ctx.scale(viewZoom, viewZoom);
        var a = canvasWidth / viewZoom,
            b = canvasHeight / viewZoom;
        for (var c = -.5 + (-nodeX + a / 2) % 50; c < a; c += 50) {
            ctx.moveTo(c, 0);
            ctx.lineTo(c, b);
        }
        ctx.stroke();
        ctx.beginPath();
        for (c = -.5 + (-nodeY + b / 2) % 50; c < b; c += 50) {
            ctx.moveTo(0, c);
            ctx.lineTo(a, c);
        }
		ctx.stroke()
        ctx.restore()
    }

**Comment by MastaCoder** [comment]
I can't check right now, can anyone confirm the above comment works? @AuroraYT

**Comment by iamtaeuk** [comment]
@MastaCoder Confirmed working. :)

**Comment by HAlexTM** [comment]
Merged #435

**Comment by AuroraYT** [comment]
No Problem :)
